proc/groupby/groupby.go

```diff
@@ -172,22 +172,29 @@ func (p *Proc) Done() {
 }
 
 func (p *Proc) run() {
+	var eof bool
 	for {
 		batch, err := p.parent.Pull()
-		if err != nil {
+		if err != nil || (batch == nil && eof) {
 			p.shutdown(err)
 			return
 		}
 		if batch == nil {
 			for {
 				b, err := p.agg.Results(true)
-				if b == nil {
+				if err != nil {
 					p.shutdown(err)
 					return
 				}
-				p.sendResult(b, err)
+				if b == nil {
+					break
+				}
+				p.sendResult(b, nil)
 			}
+			eof = true
+			continue
```

Comment:
Might be clearer to label the outer loop as PullLoop and do this.
```suggestion
				if b == nil {
					eof = true
					continue PullLoop
				}
				p.sendResult(b, nil)
			}
```

Reply:
That doesn't work because of nested loops.  It would be cleaner to put that loop in another method.

Reply:
It works if you label the outer loop.